include/tvm/relax/nested_msg.h

```diff
@@ -305,6 +305,92 @@ NestedMsg<T> MapToNestedMsg(StructInfo sinfo, FType fmapleaf) {
   }
 }
 
+/*!
+ * \brief Map expr with possible nested-tuple to nested message.
+ *
+ * This function will unpack recursive expr by its struct info and
+ * run fmapleaf for each leaf, then recursively combines the results
+ * together into a NestedMsg.
+ *
+ * The nesting structure will corresponds to the struct info of expr.
+ *
+ * \param expr The input expression which should have struct info.
+ * \param fmapleaf The mapping function for each leaf with signature
+ *             NestedMsg<T> fmap(Expr)
+ * \tparam T the content type of nested msg
+ * \tparam FType The mapping function type
+ */
+template <typename T, typename FType>
+NestedMsg<T> MapToNestedMsgBySInfo(Expr expr, FType fmapleaf) {
+  auto sinfo = GetStructInfo(expr);
+  if (auto* tuple = sinfo.as<TupleStructInfoNode>()) {
+    Array<NestedMsg<T>> res;
+    res.reserve(tuple->fields.size());
+    for (size_t i = 0; i < tuple->fields.size(); ++i) {
+      Expr field;
+      if (const auto* expr_tuple = expr.as<TupleNode>()) {
+        field = expr_tuple->fields[i];
+      } else {
+        field = TupleGetItem(expr, i);
+        UpdateStructInfo(field, tuple->fields[i]);
+      }
+      res.push_back(MapToNestedMsgBySInfo<T, FType>(field, fmapleaf));
+    }
+    return res;
+  } else {
+    return fmapleaf(expr);
+  }
+}
+
+/*!
+ * \brief Map nested message back to the expr.
+ *
+ * This function will decompose the nested message and
+ * run fmapleaf for each leaf message and get the leaf expr,
+ * then recursively combines the results as tuple expr.
+ *
+ * \param msg The input nested message.
+ * \param fmapleaf The mapping function for each leaf with signature
+ *             Expr fmap(NestedMsg<T>)
+ * \tparam T the content type of nested msg
+ * \tparam FType The mapping function type
+ */
+template <typename T, typename FType>
+Expr MapFromNestedMsg(NestedMsg<T> msg, FType fmapleaf) {
+  ICHECK(!msg.IsNull()) << "Cannot map null msg.";
+
+  if (msg.IsLeaf()) {
+    return fmapleaf(msg);
```

Comment:
* `fmapleaf` is only applied to leaves, so FType would better be “T → Expr”, and we use `fmapleaf(msg.LeafValue())` to apply. Could you update the doc and the gradient.cc side?
* I would suggest having the function name “NestedMsgToExpr” (or “MapNestedMsgToExpr”). The purpose here is to stress “the result is an Expr”.